File: features/features/session/src/main/java/com/claude/remote/features/session/SessionSwitcherViewModel.kt
package com.claude.remote.features.session

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.claude.remote.core.ssh.SshClient
import com.claude.remote.core.tmux.TmuxSession
import com.claude.remote.core.tmux.TmuxSessionManager
import com.claude.remote.core.ui.components.ConnectionState
import com.claude.remote.features.settings.SettingsRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class SessionSwitcherViewModel @Inject constructor(
    private val sshClient: SshClient,
    private val tmuxSessionManager: TmuxSessionManager,
    private val settingsRepository: SettingsRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(SessionUiState())
    val uiState: StateFlow<SessionUiState> = _uiState.asStateFlow()

    @Volatile private var loadingInProgress = false

    init {
        _uiState.update { it.copy(repoHistory = settingsRepository.getRepoHistory()) }

        viewModelScope.launch {
            sshClient.connectionState.collect { state ->
                val wasDisconnected = _uiState.value.connectionState != ConnectionState.CONNECTED
                _uiState.update { it.copy(connectionState = state, isConnecting = false) }
                if (state == ConnectionState.CONNECTED && wasDisconnected) {
                    loadSessionsAndRepos()
                }
            }
        }
        autoConnect()
    }

    fun onScreenVisible() {
        val currentState = sshClient.connectionState.value
        _uiState.update { it.copy(
            repoHistory = settingsRepository.getRepoHistory(),
            connectionState = currentState
        ) }
        if (currentState == ConnectionState.CONNECTED) {
            loadSessionsAndRepos()
        }
    }

    private fun autoConnect() {
        if (sshClient.connectionState.value == ConnectionState.CONNECTED) {
            loadSessionsAndRepos()
            return
        }
        if (settingsRepository.hasPassword()) {
            connect()
        } else {
            _uiState.update { it.copy(showPasswordPrompt = true) }
        }
    }

    fun connect() {
        val host = settingsRepository.getSshHost()
        val port = settingsRepository.getSshPort().toIntOrNull() ?: 22
        val username = settingsRepository.getSshUsername()
        val password = settingsRepository.getSshPassword()

        if (password.isEmpty()) {
            _uiState.update { it.copy(showPasswordPrompt = true) }
            return
        }

        _uiState.update { it.copy(isConnecting = true, error = null) }
        viewModelScope.launch {
            try {
                sshClient.connect(host, port, username, password)
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(error = "Connection failed: ${e.message}", isConnecting = false)
                }
            }
        }
    }

    fun connectWithPassword(password: String) {
        settingsRepository.setSshPassword(password)
        _uiState.update { it.copy(showPasswordPrompt = false) }
        connect()
    }

    fun dismissPasswordPrompt() {
        _uiState.update { it.copy(showPasswordPrompt = false) }
    }

    /** Temporarily bypass isAttachedToTmux flag for command shell operations */
    private inline fun <T> withCommandShell(block: () -> T): T {
        val wasAttached = sshClient.isAttachedToTmux
        sshClient.isAttachedToTmux = false
        return try {
            block()
        } finally {
            sshClient.isAttachedToTmux = wasAttached
        }
    }

    fun loadSessionsAndRepos() {
        if (loadingInProgress) return
        loadingInProgress = true
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val sessions = withCommandShell {
                    tmuxSessionManager.listSessions(sshClient)
                }
                _uiState.update {
                    it.copy(sessions = sessions, isLoading = false, error = null)
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        error = "Load failed: ${e.javaClass.simpleName}: ${e.message}",
                        isLoading = false
                    )
                }
            } finally {
                loadingInProgress = false
            }
        }
    }

    fun searchRepos(query: String) {
        if (query.length < 2) {
            _uiState.update { it.copy(repos = emptyList(), isSearching = false) }
            return
        }
        viewModelScope.launch {
            try {
                val output = withCommandShell {
                    sshClient.executeCommand(
                        "find ~/Developer -maxdepth 3 -type d -name .git 2>/dev/null | sed 's|/\\.git\$||' | sed 's|.*/Developer/||' | grep -i '${query.replace("'", "\\'")}' | sort | head -20"
                    )
                }
                val repos = output.lines().filter { it.isNotBlank() }
                _uiState.update { it.copy(repos = repos, isSearching = true) }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = "Search failed: ${e.message}") }
            }
        }
    }

    fun clearSearch() {
        _uiState.update { it.copy(repos = emptyList(), isSearching = false) }
    }

    fun attachSession(sessionName: String) {
        _uiState.update { it.copy(repos = emptyList(), isSearching = false) }
        sshClient.currentSessionName = sessionName
        _uiState.update { it.copy(navigateToSession = sessionName) }
    }

    fun onNavigated() {
        _uiState.update { it.copy(navigateToSession = null) }
    }

    fun createSessionFromRepo(repo: String) {
        val sessionName = repo.substringAfterLast("/")
        val workDir = "\$HOME/Developer/$repo"
        settingsRepository.addRepoToHistory(repo)
        _uiState.update { it.copy(repos = emptyList(), isSearching = false, repoHistory = settingsRepository.getRepoHistory()) }
        viewModelScope.launch {
            try {
                withCommandShell {
                    tmuxSessionManager.createSession(sessionName, workDir, sshClient)
                }
                kotlinx.coroutines.delay(500)
                sshClient.currentSessionName = sessionName
                _uiState.update { it.copy(navigateToSession = sessionName) }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun killSession(sessionName: String) {
        viewModelScope.launch {
            try {
                withCommandShell {
                    tmuxSessionManager.killSession(sessionName, sshClient)
                }
                loadSessionsAndRepos()
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
}

File: ios/ClaudeRemote/ClaudeRemote/Features/Session/SessionSwitcherViewModel.swift
import SwiftUI
import Combine

/// Ports Android `SessionSwitcherViewModel.kt`.
@MainActor
final class SessionSwitcherViewModel: ObservableObject {
    private let ssh = SshClient.shared
    private let settings = SettingsStore.shared

    @Published var connectionState: ConnectionState = .disconnected
    @Published var isConnecting = false
    @Published var sessions: [TmuxSession] = []
    /// Live sessions grouped by project (cwd), mirroring the webmux sidebar.
    var projects: [TmuxProject] { sessions.groupedByProject() }
    @Published var repos: [String] = []
    @Published var repoHistory: [String] = []
    @Published var isSearching = false
    @Published var isLoading = false
    @Published var error: String?
    @Published var showPasswordPrompt = false
    @Published var navigateToSession: String?

    private var cancellables = Set<AnyCancellable>()
    private var loading = false

    init() {
        repoHistory = settings.getRepoHistory()
        ssh.$connectionState
            .receive(on: RunLoop.main)
            .sink { [weak self] state in
                guard let self else { return }
                let wasDisconnected = self.connectionState != .connected
                self.connectionState = state
                self.isConnecting = false
                if state == .connected && wasDisconnected { self.loadSessions() }
            }
            .store(in: &cancellables)
        autoConnect()
    }

    func onScreenVisible() {
        repoHistory = settings.getRepoHistory()
        connectionState = ssh.connectionState
        if ssh.connectionState == .connected { loadSessions() }
    }

    private func autoConnect() {
        if ssh.connectionState == .connected { loadSessions(); return }
        if settings.hasPassword() { connect() } else { showPasswordPrompt = true }
    }

    func connect() {
        let pw = settings.getSshPassword()
        if pw.isEmpty { showPasswordPrompt = true; return }
        isConnecting = true; error = nil
        Task {
            do {
                try await ssh.connect(host: settings.sshHost,
                                      port: Int(settings.sshPort) ?? 22,
                                      username: settings.sshUsername,
                                      password: pw)
            } catch {
                self.error = "Connection failed: \(error.localizedDescription)"
                isConnecting = false
            }
        }
    }

    func connectWithPassword(_ pw: String) {
        settings.setSshPassword(pw)
        showPasswordPrompt = false
        connect()
    }

    func dismissPasswordPrompt() { showPasswordPrompt = false }

    func loadSessions() {
        if loading { return }
        loading = true
        Task {
            isLoading = true; error = nil
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            do {
                sessions = try await TmuxSessionManager.listSessions(ssh)
                isLoading = false
            } catch {
                self.error = "Load failed: \(error.localizedDescription)"
                isLoading = false
            }
            ssh.isAttachedToTmux = wasAttached
            loading = false
        }
    }

    func searchRepos(_ query: String) {
        if query.count < 2 { repos = []; isSearching = false; return }
        Task {
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            do {
                repos = try await TmuxSessionManager.searchRepos(query: query, ssh)
                isSearching = true
            } catch { self.error = "Search failed: \(error.localizedDescription)" }
            ssh.isAttachedToTmux = wasAttached
        }
    }

    func attachSession(_ name: String) {
        repos = []; isSearching = false
        ssh.currentSessionName = name
        navigateToSession = name
    }

    func onNavigated() { navigateToSession = nil }

    func createSessionFromRepo(_ repo: String) {
        let name = String(repo.split(separator: "/").last ?? Substring(repo))
        let workDir = "$HOME/Developer/\(repo)"
        settings.addRepoToHistory(repo)
        repos = []; isSearching = false; repoHistory = settings.getRepoHistory()
        Task {
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            do {
                try await TmuxSessionManager.createSession(name: name, workDir: workDir, ssh)
                try? await Task.sleep(for: .milliseconds(500))
                ssh.isAttachedToTmux = wasAttached
                ssh.currentSessionName = name
                navigateToSession = name
            } catch {
                ssh.isAttachedToTmux = wasAttached
                self.error = error.localizedDescription
            }
        }
    }

    func killSession(_ name: String) {
        Task {
            let wasAttached = ssh.isAttachedToTmux
            ssh.isAttachedToTmux = false
            do {
                try await TmuxSessionManager.killSession(name, ssh)
                ssh.isAttachedToTmux = wasAttached
                loadSessions()
            } catch {
                ssh.isAttachedToTmux = wasAttached
                self.error = error.localizedDescription
            }
        }
    }

    func clearError() { error = nil }
}
